### PR TITLE
Make changes nescessary to build IL asm grammar with Bison.

### PR DESCRIPTION
**Pull request by kant2002**
To generate with Bison
```
cd src/coreclr/ilasm
yacc asmparse.y --output=prebuilt/asmparse.cpp
```

If bison is unacceptable due to licensing
```
cd src/coreclr/ilasm
byacc asmparse.y --output=prebuilt/asmparse.cpp
```

Thanks @hez2010 for the hint.

Related to #4776

### CODE REVIEW

**Comment by kant2002** [comment]
Would be great if my changes was tested with internal yacc, so while build infra prepare nescessary images and stuff we can test this little experiment

**Comment by RIP-webmaster** [comment]
Unrelated to this change, but do someone know where "Lexical tokens" entries in asmparse.grammar come from? Looks like some of them are not accurate, for example: `ID - C style alphaNumeric identifier (e.g. Hello_There2)` - IlAsm identifiers can also contain character "\`" (backtick), while C-style identifiers can only contain letters, digits and underscores. Where the exact meaning of tokens is specified?

File: src/coreclr/ilasm/asmparse.y [hunk 1]
```diff
@@ -47,8 +47,8 @@
 %token <string>  DOTTEDNAME     /* System.Object */
 %token <binstr>  QSTRING        /* "Hello World\n" */
 %token <string>  SQSTRING       /* 'Hello World\n' */
-%token <int32>   INT32          /* 3425 0x34FA  0352  */
-%token <int64>   INT64          /* 342534523534534      0x34FA434644554 */
+%token <int32>   INT32_V        /* 3425 0x34FA  0352  */
+%token <int64>   INT64_V        /* 342534523534534      0x34FA434644554 */
```
Comment (hez2010) [inline src/coreclr/ilasm/asmparse.y:51]: I would instead recommend `INT32_T` and `INT64_T` for naming.

File: src/coreclr/ilasm/asmparse.h [hunk 2]
```diff
@@ -314,6 +314,7 @@ class AsmParse : public ErrorReporter
     friend char* nextBlank(_In_ __nullterminated char*);
     friend int ProcessEOF();
     friend unsigned __int8* skipType(unsigned __int8* ptr, BOOL fFixupType);
+    friend unsigned corCountArgs(BinStr* args);
```
Comment (hez2010) [inline src/coreclr/ilasm/asmparse.h:317]: This is unnecessary. 
Instead, add a `static unsigned corCountArgs(BinStr* args);` declaration in grammar_before.cpp.

File: src/coreclr/ilasm/grammar_after.cpp [hunk 3]
```diff
@@ -1557,7 +1557,7 @@ void FixupTyPars(BinStr* pbstype)
     FixupTyPars((PCOR_SIGNATURE)(pbstype->ptr()),(ULONG)(pbstype->length()));
 }
 /**************************************************************************/
-static unsigned corCountArgs(BinStr* args)
+unsigned corCountArgs(BinStr* args)
```
Comment (hez2010) [inline src/coreclr/ilasm/grammar_after.cpp:1560]: Unnecessary change.

**Review by hez2010 — CHANGES_REQUESTED** [review]

File: src/coreclr/ilasm/asmparse.y (hunk 1 above)
Comment (kant2002) [reply src/coreclr/ilasm/asmparse.y:51]: I assume _T is token here. I use _V for value. but whatether, let maintainers decide.

**Comment by kant2002** [comment]
@RIP-webmaster for example how numbers as lexems are parsed is defined here.
https://github.com/dotnet/runtime/blob/242242947c01615b243e8b4d9ee163030a4ceb47/src/coreclr/ilasm/grammar_after.cpp#L1184-L1229

Identifier rules 
https://github.com/dotnet/runtime/blob/242242947c01615b243e8b4d9ee163030a4ceb47/src/coreclr/ilasm/grammar_after.cpp#L548-L558

which is used here https://github.com/dotnet/runtime/blob/242242947c01615b243e8b4d9ee163030a4ceb47/src/coreclr/ilasm/grammar_after.cpp#L903-L918

all of that pretty standard lexer stuff encoded in the code directly.

File: src/coreclr/ilasm/asmparse.y (hunk 1 above)
Comment (hamarb123) [reply src/coreclr/ilasm/asmparse.y:51]: My 2 cents are: I personally like `_V` for value. If it were `_T` I would expect a token representing the type, not a value.

File: src/coreclr/ilasm/grammar_before.cpp [hunk 4]
```diff
@@ -51,6 +51,7 @@ static char* newString(_In_ __nullterminated const char* str1);
 static void corEmitInt(BinStr* buff, unsigned data);
 static void AppendStringWithLength(BinStr* pbs, _In_ __nullterminated char* sz);
 static void AppendFieldToCustomBlob(BinStr* pBlob, _In_ BinStr* pField);
+static unsigned corCountArgs(BinStr* args);
```
Comment (hez2010) [inline src/coreclr/ilasm/grammar_before.cpp:54]: Also append below lines after this line:
```cpp
extern void yyerror(_In_ __nullterminated const char*);
extern Instr* SetupInstr(unsigned short);
extern int yylex();
```

Comment (kant2002) [reply src/coreclr/ilasm/grammar_before.cpp:54]: Can you clarify why extern needed?

**Review by hez2010 — APPROVED** [review]
LGTM

File: src/coreclr/ilasm/asmparse.y (hunk 1 above)
Comment (hez2010) [inline src/coreclr/ilasm/asmparse.y:51]: We don't have to change it if we use byacc instead of bison.
And worth to note by using bison there will be license conflict between GPL and MIT. So I would recommend to revert this change and use byacc instead.

**Comment by JulieLeeMSFT** [comment]
@TIHan please review this community PR.

File: src/coreclr/ilasm/asmparse.y (hunk 1 above)
Comment (TIHan) [reply src/coreclr/ilasm/asmparse.y:51]: "_V" is fine.

Comment (TIHan) [reply src/coreclr/ilasm/asmparse.y:51]: [BISON Conditions](https://www.gnu.org/software/bison/manual/html_node/Conditions.html)

I'm not a lawyer. @richlander , what do you think? This only applies to the parser the BISON generates from the grammar.

**Review by TIHan — CHANGES_REQUESTED** [review]
Overall, this looks like a very simple change, though I do not know what the implications are with BISON being GPL and what the means for the parsers that are generated.

We need to get some insight with @richlander .

**Comment by TIHan** [comment]
@kant2002 , we are going to check to see if the generated parser from BISON will be license compatible. If it isn't, will `byacc` work?

File: src/coreclr/ilasm/asmparse.y (hunk 1 above)
Comment (hez2010) [reply src/coreclr/ilasm/asmparse.y:51]: Reverting this can help prevent developers from generating code using bison by accident (because it would just fail), which can make sure we won't have license issues.

**Comment by TIHan** [comment]
Re-opening, we can use this as long as we don't remove the special exception that BISON generates, very important.

**Comment by TIHan** [comment]
@kant2002 we are willing to accept this. Can you generate the new parser and make it part of this PR?

**Comment by TIHan** [comment]
No worries and no rush.

File: src/coreclr/ilasm/asmparse.y (hunk 1 above)
Comment (hamarb123) [reply src/coreclr/ilasm/asmparse.y:51]: I think it would be better to use byacc due to the licensing personally.

**Comment by JulieLeeMSFT** [comment]
@kant2002, this is a kindly reminder. 

> @kant2002 we are willing to accept this. Can you generate the new parser and make it part of this PR?
